README
SQL/AnyDBD version 0.03
=======================

See POD for documentation.

INSTALLATION

To install this module type the following:

   perl -e 'install SQL::AnyDBD;'

or

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  DBIx::AnyDBD
  Params::Validate

COPYRIGHT AND LICENCE

Copyright (C) 2003 U-MOKSHA\metaperl

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

Changes
Revision history for Perl extension SQL::AnyDBD.

0.03  Sat Jul  7 18:35:11 2007
	- DMUEY contributed:
		- removed unnecessary Exporter use from AnyDBD.pm 
		- streamlined use of Params::Validate and what it [ex|im]ported
	    - fixed regexes used in Params::Validate checks
		- Added and documented AS(), function(), count(), get_sql_from_files(), sql_driver_version(), sql_driver_name(), normalized_sql_driver_name(), get_placeholder_string(), and get_placeholder_array()
		- Generated SQL should be a bit cleaner (join only defined values, etc)
		- SELECT not require "tables" so you could do a "SELECT NOW();" type query
		- Cleaned up Makefile.PL including removing restriction to 5.8
		- Cleaned up README
		- General Code cleanup as per "Perl Best Practices"
		- redo tests to use pseudo sqlite so no preset info/prompts/db driver are required
		- [todo] create and include SQL::AnyDBD::Oracle.pm, change 'big 3' refs to include changes [/todo]
		
0.01  Tue Dec 30 20:34:52 2003
	- original version; created by h2xs 1.22 with options
		-A -X -n SQL::AnyDBD



lib/SQL/AnyDBD/Default.pm
package SQL::AnyDBD::Default;

use strict;
use Params::Validate qw(:types validate_with);

sub SELECT {
    my $sa = shift;
    
    my %p = validate_with(
        'params' => \@_, 
        'spec'   => { 
		    'fields'   => { 
		        'type'     => ARRAYREF,  
		        'optional' => 0, 
		    },
		    'tables'   => { 
		        'type'     => ARRAYREF,  
		        'optional' => 1, 
		    },
		    'where'    => { 
		        'type'     => SCALAR,    
		        'optional' => 1, 
		    },
		    'group_by' => { 
		        'type'     => SCALAR,    
		        'optional' => 1, 
		    },
		    'having'   => { 
		        'type'     => SCALAR,    
		        'optional' => 1, 
		    },
		    'order_by' => { 
		        'type'     => SCALAR,    
		        'optional' => 1, 
		    },
		    'limit'    => { 
		        'type'     => HASHREF,   
		        'optional' => 1, 
		    },
		},
	);

    my $fields   = join ',', @{ $p{'fields'} };
    my $tables   = $p{'tables'}   ? 'FROM ' . join(',', @{ $p{'tables'} }) : undef;
    my $where    = $p{'where'}    ? "WHERE $p{'where'}"                    : undef;
    my $group_by = $p{'group_by'} ? "GROUP BY $p{'group_by'}"              : undef;
    my $having   = $p{'having'}   ? "HAVING $p{'having'}"                  : undef;
    my $order_by = $p{'order_by'} ? "ORDER BY $p{'order_by'}"              : undef;
    my $limit    
        = ( scalar keys %{ $p{'limit'} } ) ? $sa->LIMIT( %{ $p{'limit'} } ) : undef;

    return 'SELECT ' . join(' ', grep { defined } $fields, $tables, $where, $group_by, $having, $order_by, $limit);
}

sub UPDATE {
    my $sa = shift;
    
    my %p = validate_with(
        'params' => \@_, 
        'spec'   => { 
		    'table' => { 
		        'type'     => SCALAR,  
		        'optional' => 0, 
		    },
		    'set'   => { 
		        'type'     => SCALAR,  
		        'optional' => 0, 
		    },
		    'where' => { 
		        'type'     => SCALAR,  
		        'optional' => 1, 
		    },
		    'limit' => { 
		        'type'     => SCALAR,  
		        'optional' => 1,
		    },
		},
    );
    
    my $where = $p{'where'} ? "WHERE $p{'where'}" : undef;
    my $limit = $p{'limit'} ? "LIMIT $p{'limit'}" : undef;

    return 'UPDATE ' . join(' ', grep { defined } $p{'table'}, 'SET', $p{'set'}, $where, $limit);
}

sub INSERT {
    my $sa = shift;
    
    my %p = validate_with(
        'params' => \@_, 
        'spec'   => { 
		    'table'   => { 
		        'type'     => SCALAR,    
		        'optional' => 0, 
		    },
		    'columns' => { 
		        'type'     => ARRAYREF,  
		        'optional' => 1, 
		    },
		    'values' => { 
		        'type'     => ARRAYREF,  
		        'optional' => 0, 
		    },
        },
    );

    my $columns = scalar @{ $p{'columns'} } 
        ? sprintf '(%s)', join( ',', @{ $p{'columns'} } ) : undef;
    my $values  = sprintf '(%s)', join( ',', @{ $p{'values'} } );

    return 'INSERT INTO ' . join(' ', grep { defined } $p{'table'}, $columns, 'VALUES', $values);
}

sub DELETE {
    my $sa = shift;
    
    my %p = validate_with(
        'params' => \@_, 
        'spec'   => { 
		    'table'    => { 
		        'type'     => SCALAR, 
		        'optional' => 0, 
		    },
		    'where'    => { 
		        'type'     => SCALAR, 
		        'optional' => 1, 
		    },
		    'group_by' => { 
		        'type'     => SCALAR, 
		        'optional' => 1, 
		    },
		    'order_by' => { 
		        'type'     => SCALAR, 
		        'optional' => 1, 
		    },
		    'limit'    => { 
		        'type'     => SCALAR, 
		        'optional' => 1, 
		    },
        },
    );

    my $where    = $p{'where'}    ? "WHERE $p{'where'}"       : undef;
    my $group_by = $p{'group_by'} ? "GROUP BY $p{'group_by'}" : undef;
    my $order_by = $p{'order_by'} ? "ORDER BY $p{'order_by'}" : undef;
    my $limit    = $p{'limit'}    ? "LIMIT $p{'limit'}"       : undef;

    return 'DELETE FROM ' . join(' ', grep { defined }  $p{'table'}, $where, $group_by, $order_by, $limit);
}

sub IN {
    my $sa = shift;
   
    my %p = validate_with(
        'params' => \@_, 
        'spec'   => { 
            'values' => { 
                'type'     => ARRAYREF, 
                'optional' => 0,
            },
		},
    );
   
    my $values = join ',', @{ $p{'values'} };
    
    return $values ? "IN ($values)" : '';
}

sub AS {
    my ($sa, $real, $alias) = @_;
    return "$real AS $alias";
}

# General Utility methods:

sub sql_driver_version {
    return shift->VERSION() || $SQL::AnyDBD::VERSION;    
}

sub get_placeholder_string {
    shift;
    return join( ',', split '', ( '?' x (ref $_[0] eq 'ARRAY' ? @{ $_[0] } : @_) ) );
}

sub get_placeholder_array {
    shift;
    return wantarray ?   map { '?' } (ref $_[0] eq 'ARRAY' ? @{ $_[0] } : @_) 
                     : [ map { '?' } (ref $_[0] eq 'ARRAY' ? @{ $_[0] } : @_) ]
                     ;
}

sub function {
    my ($sa, $function, $arg_str) = @_;
    $function = uc $function if !$sa->{'no_uc'};
    $arg_str  = '' if !defined $arg_str;
    return "$function($arg_str)";
}

sub get_sql_from_files {
    my ($sa, @files, $pre_process) = @_;
    
    my $slurp_sql = '';
    
    if( ref $pre_process ne 'CODE' ) {
        push @files, $pre_process;
        $pre_process = '';
    }
    
    for my $file (@files) {
        
        my $real = $file;
        $real   .=  '.' . $sa->normalized_sql_driver_name() 
            if -e $real . '.' . $sa->normalized_sql_driver_name();
        
        if( open my $sql_fh, '<', $real ) {
            my $sql = do { local $/; <$sql_fh> };
            close $sql_fh;
        
            if( ref $pre_process eq 'CODE' ) {
                $slurp_sql .= $pre_process->( $sa, $sql, $file, $real );
            }
            else {
                $slurp_sql .= $sql;
            }
        }
        else {
            return;
        }
    }
    
    return $slurp_sql;
}

sub sql_driver_name {
    return (reverse split "::", ref shift)[0];
}

sub normalized_sql_driver_name {
    return lc( shift->sql_driver_name() );
}

sub _init {
    my ($sa) = @_;
    $sa->{'no_uc'} = 0 if !defined $sa->{'no_uc'};
}

1;

lib/SQL/AnyDBD.pm
package SQL::AnyDBD;

use strict;
use warnings;
use base 'DBIx::AnyDBD';

our $VERSION = '0.03';

sub new {
    my ($pkg, $dbh) = @_;
    my $self = bless { 'package' => __PACKAGE__ , dbh => $dbh }, __PACKAGE__;
    $self->rebless;
    $self->_init if $self->can('_init');
    return $self; 
}

1;

__END__

=head1 NAME

SQL::AnyDBD - Perl extension to generate SQL for RDBMS variants

=head1 SYNOPSIS

  use SQL::AnyDBD;

  my $dbh = DBI->connect($dsn, $user, $pass, $attr) or die $!;
  my $sa  = SQL::AnyDBD->new($dbh);

  my $rows_desired = 8;
  my $start_row    = 4;

  warn $sa->LIMIT(rows_desired => $rows_desired, start_row => $start_row);

  # yields ...

  LIMIT 8 OFFSET 4  -- for Pg and SQLite
  4,8               -- for Mysql


=head1 ABSTRACT

SQL::AnyDBD is module which generates SQL for different RDBMSes from a 
uniform API. It uses DBIx::AnyDBD to determine which SQL variant to
generate. In this documentation, "the big 3" is used to refer to the 3 free
popular databases: Postgresql, SQLite, and MySQL (There is a reason I listed
them in that order, but now is no time to get into a holy war. :)


=head1 METHODS

=head2 $sa->LIMIT(rows_desired => $r [ , start_row => $s ] )

=over 4

=item REQUIRED: rows_desired

=item OPTIONAL: start_row

=back

A limit clause is used to limit the result set from an SQL select. 
Each of the big 3 supports this concept. All 3 also accept integers for both
arguments. However, Pg can also accept the term C<ALL> for C<rows_desired>.  

=head2 $sa->IN ( values => \@v )

=over 4 

=item REQUIRED: values

=back

It is required that the arrayref C<values> be passed. 
It is not required that the
arrayref have any data:

 my $sql = $sa->IN(values => [12..14]);

 # yields...

 IN (12,13,14)

=head2 $sa->AS ( $real, $alias)

=over 4 

=item REQUIRED: both arguments

=back

 my $sql = $sa->AS('userprefs', 'up');
 
 # yields...

 userprefs AS up

=head2 $sa->SELECT ( fields => \@f, tables => \@t, ... )

=over 4 

=item REQUIRED: fields, 

=item OPTIONAL: tables, where, group_by, having, order_by, limit

=back

All the optional arguments take a string as an argument with the exception of
C<limit> which takes a hashref which is passed to C<LIMIT()>. 'tables' and the required
arguments take an arrayref  B<even if they only have one argument>: 

 my %select = 
	(
	 fields   => [qw(student_ssan)],
	 tables   => ["student INNER JOIN classes"],
	 where    => "student.classes_id = classes.classes_id",
	 group_by => "classes_year",
	 having   => "student_age > 30",
	 order_by => 'student_id',
	 limit    => { rows_desired => 5, start_row => 77 },
	);

 my $sa  = SQL::AnyDBD->new($dbh);
 my $sql = $sa->SELECT(%select);

 # yields...

 SELECT 
    student_ssan 
 FROM 
    student INNER JOIN classes 
 WHERE 
    student.classes_id = classes.classes_id 
 GROUP BY 
    classes_year 
 HAVING 
    student_age > 30 
 ORDER BY 
    student_id 
 LIMIT  
    5 
 OFFSET 
    77

As of 0.03 'tables' is optional so that you can do queries liek this:
 
 my $sql = $sa->SELECT( 'fields' => [ $sa->function('NOW', time) ]);
  
 #yields:
 
 SELECT NOW(1170616724)

=head2 $sa->DELETE ( table => $tbl [ ,  ...  ] )

=over 4 

=item REQUIRED: table

=item OPTIONAL:  where, order_by, limit

=back

C<table> is the name of the table from which to delete records. C<where> is a
string specifying the filtering of rows. C<order_by> is useful in
conjunction with C<limit> in order to delete rows based on order:

 my %parms = 
	(
	 table => 'student',
	 where  => "student.classes_id = 420",
	 order_by => 'date_enrolled',
	 limit  => 4
	);

 my $sa  = SQL::AnyDBD->new($dbh);
 my $sql = $sa->DELETE(%parms);

 # yields ...

 DELETE FROM 
   student 
 WHERE 
   student.classes_id = 420 
 ORDER BY 
   date_enrolled 
 LIMIT 4

=head2 $sa->UPDATE ( set => $set_expr, , table => $t [, ...] )

=over 4 

=item REQUIRED: set, table

=item OPTIONAL: where, limit

=back

The required argument set is a string consisting of a series of

   col_name1=expr1 [, col_name2=expr2, ...]

expressions. The required argument C<table> is a table name.


All the optional arguments take a scalar as an argument.
C<limit> which takes a hashref which is passed to C<LIMIT()>. Both required
arguments take an arrayref  B<even if they only have one argument>:

 my %update =
	(
         table  => "student",
	 set    => "student_ssan = NULL",
	 where  => "student_country_id <> 1",
	 limit  => 12
	);

 my $sa  = SQL::AnyDBD->new($dbh);
 my $sql = $sa->UPDATE(%update);

 # yields...

 UPDATE student SET student_ssan = NULL WHERE student_country_id <> 1 LIMIT 12

=head2 $sa->INSERT ( table => $tbl, values => \@values ...

=over 4 

=item REQUIRED: table, values

=item OPTIONAL: columns

=back

 my %insert =
   (
     table     => "student",
     columns   => [qw(student_ssan3 student_ssan2 student_ssan4 
                      student_lname student_fname)],
     values    => [qw(123 45 9876 olajuwon hakeem)]
   );

 my $sa  = SQL::AnyDBD->new($dbh);
 my $sql = $sa->INSERT(%insert);

 # yields ...

 INSERT INTO student 
   (student_ssan3, student_ssan2, student_ssan4,student_lname,student_fname) 
 VALUES
   (123,           45,            9876,         olajuwon,     hakeem)

=head2 Utility methods

=head3 $sa->function($function_name, $argument_string)

This utility method helps us to consistently render funtion call SQL.

    $sa->function('foo') # FOO()
    $sa->function('bar', q{'baz = 1'}) # BAR('baz = 1')

A subclass might have a lookup hash in its function() where it translates common functions to its specific type and set default arguments:

    $sa->function('NOW') # SYSTEM_TIME( GMTOFFSET -4 )

=head3 $sa->get_placeholder_string()

Pass a single array ref or array and you get back a string suitable for placeholding ('?' joined with commas):

   my @people = (
       ['Larry', 'Wall'],
       ['Homer', 'Simpson'],
       ['Joe', 'Mama'],
   );

   my $sth = $dbh->prepare( 
       $sa->INSERT(
          'table'  => 'people',
          'values' => [ 'NULL', $sa->get_placeholder_string( $people[0] ) ], # ?,?
       )
   );
   # INSERT INTO people VALUES(NULL,?,?)
   
   for my $person ( @people ) {
       $sth->execute( @{ $person } );
   }

=head3 $sa->get_placeholder_array()

Same as get_placeholder_string() except it returns an array of '?' placeholder
(or an array ref of the same in scalar context)

=head3 $sa->get_sql_from_files()

Takes a list of files and concatenates their contents together.

If a given argument exists with a dot-normalized_sql_driver_name that is used instead.

If you are using SQLite and do this:

   my $sql = $sa->get_sql_from_files('./sql/schema');

Then is uses ./sql/schema.sqlite if it exists, and ./sql/schema if it does not.

If the last argument is a code ref its return value is what gets concatenated instead of the raw file.
It gets called with these arguments:
  SQL::AnyDBD-object, SQL,       FILE_PASSED,    FILE_OPENED
  $sa,               -contents-, './sql/schema', './sql/schema.sqlite'

=head3 $sa->normalized_sql_driver_name()

Normalized version of $sa->sql_driver_name()

=head3 $sa->sql_driver_name()

SQL::AnyDBD driver name

=head3 $sa->sql_driver_version()

SQL::AnyDBD driver version number.

=head3 $sa->get_dbh()

Returns the $dbh passed to new(). provided by L<DBIx::AnyDBD>

=head1 SEE ALSO

L<DBIx::Std>

=head2 Multi-database Products on CPAN

=over 4

=item L<DBIx::Std>

=item L<Rose::DB>

=item L<SQL::Translator>

=item L<Alzabo>

=item L<Class::DBI>

=item L<DBIx::Recordset>

=item L<DBIx::AnyDBD>

=back

=head2 Supported RDBMSes

=over 4

=item Postgresql

http://www.postgresql.org

=item SQLite

http://www.sqlite.org

=item MySQL

http://www.mysql.com

=back

=head1 AUTHOR

Terrence Brannon, <tbone@cpan.org>

v0.03 Co-maintainer Daniel Muey, L<http://drmuey.com/cpan_contact.pl>

=head1 COPYRIGHT AND LICENSE

Copyright 2003 by Terrence Brannon

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut
